### ISSUE TITLE
Hostfolder configuration should not save the password in clear text

### ISSUE THREAD

**Issue by LalatenduMohanty** [minishift/minishift#661]
minishift hostfolder saves the password in clear text. So anyone who will get the access of  the workstation can view the file and can get read the password
```
$ cat ~/.minishift/config/allinstances.json 
{
	"HostFolders": [
		{
			"Name": "share",
			"Type": "cifs",
			"Options": {
				"domain": "test",
				"mountpoint": "",
				"password": "sample-password",
				"uncpath": "/Users/lmohanty/",
				"username": "lmohanty"
			}
		}
	]
```

**Comment by LalatenduMohanty** [comment]
@hferentschik Also mentioned this and I saw this while I was trying `minishift hostfolder`

**Comment by gbraad** [comment]
@LalatenduMohanty the same behaviour is known for Docker4Windows. Encryption on the workstation, and decrypting is no security.

**Comment by hferentschik** [comment]
So here are some of the options I see atm. As @gbraad is saying there is no way to get rid of this for CIFS. the sshfs approach is double without passwords and might be for that reason in the long run the the better option. Anyways:

* Have a mode which does not store the password. It interactively asks for the password on each start. Independent of what else we are going to do, I think this option makes sense for very password cautious people. You basically get prompted for the password during start.
* Instead of storing them clear text we can scramble them. Obviously this is not really adding security per se, but it is at least a bit better imo. At least one needs to know/extract what Minishift applies to retrieve the plain password
* Implement some sort of password vault. Obviously this a lot of work and the user still needs a master password. 

ATM, I tend to think a combination of option one and two might work. Option one for people who really don't want to have the password stored anywhere and option two to make it at least a bit harder to get to the password.

> Storing a credentials file inside the VM might work on instance level, but for all instances, we would still need to store credentials and push them to the VM.

Right.

**Comment by gbraad** [comment]
This is what we had discussed, yes...

> Instead of storing them clear text we can scramble them.

Security through obscurity is no security ;-)... is like applying ROT13, just to make it slightly less easy to guess at first sight,

> It interactively asks for the password on each start. 

> Implement some sort of password vault. 

These are the better options. This allows the user to choose it their own security approach. Especially the first feels much like having a password-protected certificate. And the user COULD allow to remove the security, but chooses in this case to have cleartext being stored.

I prefer actually option 1... with the user deciding to do clear text storing as 4th possible option.

**Comment by LalatenduMohanty** [comment]
> the sshfs approach is double without passwords and might be for that reason in the long run the the better option.

Another reason to have host folder sharing using sshfs. 

> Have a mode which does not store the password. It interactively asks for the password on each start. Independent of what else we are going to do, I think this option makes sense for very password cautious people. You basically get prompted for the password during start.

+1. I would prefer this one. 

> Instead of storing them clear text we can scramble them. Obviously this is not really adding security per se, but it is at least a bit better imo. At least one needs to know/extract what Minishift applies to retrieve the plain password

This also works. Basically the username and password get stored  in another file and get referenced .

**Comment by LalatenduMohanty** [comment]
@gbraad I assuming we would have both kind of users some who wont care if their password is stored in clear text and some would me mad about this. Wondering how DockerForWindows did not get bad flack about this.

**Comment by gbraad** [comment]
@LalatenduMohanty the situation for D4W is slightly more complicated than explained here. they do utilize a native store option, But still, the keys are stored on the account of the user and retrievable when the account is logged in.

**Comment by hferentschik** [comment]
Merged via pull request #762
